Cling: prevent double release of Transactions

smuzaffar opened the pull request.

Axel-Naumann committed:
[cling] Fix alignment of Transaction allocation:

new of a char array might not have the correct alignment to hold a Transaction.
Allocate a Transaction itself directly, instead of in-place constructing it in
the character array.

Each Transaction in the pool is thus constructed through `new Transaction(S)`
and destructed through `delete T`, which is nicely symmetrical. The use of
`::operator new` and `::operator delete` isn't actually necessary.

While I'm at it, improve the assert message's wording.

Axel-Naumann committed:
[cling] Prevent double delete / double TransactionPool-ing:

When a Transaction is unloaded for which *also* a ScopedTransactionRAII
is waiting, the latter will potentially access a deleted Transaction.
It can be deleted due to the Pool being full. Even if it is not deleted
(as is the case in https://github.com/root-project/root/issues/7657 )
it will get pushed into the Pool once by unload() and a second time by
the RAII!

The two options to track this case were full-blown ref counting or noting
that a Transaction is attached to an RAII scope. If that is the case,
freeing the Transaction must be skipped, as the RAII will take care of it.

Axel-Naumann committed:
[cling] Remove historical leftover in TransactionPool size (NFC).

Axel-Naumann committed:
[cling] Assert released transaction is != last in pool:

a cheap way to notice what went wrong in https://github.com/root-project/root/issues/7657.

smuzaffar commented:
please test with cms-sw/cmsdist#6778

cmsbuild commented:
A new Pull Request was created by  @smuzaffar (Malik Shahzad Muzaffar) for branch cms/master/f078957.

@smuzaffar, @mrodozov can you please review it and eventually sign? Thanks.
cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

cmsbuild commented:
-1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-56e843/13892/summary.html
**COMMIT**: 734f63c971e7567ce0efc78d883b31d2592adcd7
**CMSSW**: CMSSW_11_3_ROOT6_X_2021-03-30-2300/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/root/152/13892/install.sh` to create a dev area with all the needed externals and cmssw changes.


## External Build

I found compilation error when building: 

<pre>make[2]: *** Waiting for unfinished jobs....
make[2]: Leaving directory `/data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/BUILD/slc7_amd64_gcc900/external/dd4hep/v01-16x-789ded0f3abbae7c552fa6d7a7d9bd76/build'
make[1]: *** [DDCore/CMakeFiles/DDCore.dir/all] Error 2
make[1]: Leaving directory `/data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/BUILD/slc7_amd64_gcc900/external/dd4hep/v01-16x-789ded0f3abbae7c552fa6d7a7d9bd76/build'
make: *** [all] Error 2
error: Bad exit status from /data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/tmp/rpm-tmp.E0wJ12 (%build)


RPM build errors:
Macro %rpmbuild_libdir defined but not used within scope
Bad exit status from /data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/tmp/rpm-tmp.E0wJ12 (%build)
</pre>

smuzaffar commented:
please test with cms-sw/cmsdist#6778

Axel-Naumann committed:
[cling] Move PushTransactionRAII out of Interpreter:

This will allow Transaction remembering its RAII to not depend on Interpreter.

Axel-Naumann committed:
[cling] Order includes alphabetically (NFC).

Axel-Naumann committed:
[cling] Transaction needs to know its RAII:

This will allow the Intrepreter to prevent unload() on
Transactions held by RAIIs.

Axel-Naumann committed:
[cling] Assert that no nested attaches to unloading Transaction.

cmsbuild commented:
Pull request #152 was updated.

smuzaffar commented:
please test with cms-sw/cmsdist#6778

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-56e843/13901/summary.html
**COMMIT**: b9fa1a4395b501ed9784773b702e17864b05eae8
**CMSSW**: CMSSW_11_3_ROOT6_X_2021-03-30-2300/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/root/152/13901/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 15 differences found in the comparisons
 - DQMHistoTests: Total files compared: 37
 - DQMHistoTests: Total histograms compared: 2640841
 - DQMHistoTests: Total failures: 13
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2640806
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 36 files compared)
 - Checked 155 log files, 37 edm output root files, 37 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_11_3_ROOT6_X_2021-03-30-2300+56e843/42002/triggerResults): no differences found

smuzaffar commented:
please test with cms-sw/cmsdist#6778

smuzaffar commented:
test parameters:
- disable_poison = true

smuzaffar commented:
please test with cms-sw/cmsdist#6778

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-56e843/13916/summary.html
**COMMIT**: b9fa1a4395b501ed9784773b702e17864b05eae8
**CMSSW**: CMSSW_11_3_ROOT6_X_2021-03-31-2300/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/root/152/13916/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 7 differences found in the comparisons
 - DQMHistoTests: Total files compared: 37
 - DQMHistoTests: Total histograms compared: 2640841
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2640818
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 36 files compared)
 - Checked 155 log files, 37 edm output root files, 37 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_11_3_ROOT6_X_2021-03-31-2300+56e843/42015/triggerResults): no differences found
